ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+		if (!is_checkpoint && header.extra_data().len() != (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && header.extra_data().len() <= (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH))
+			|| (is_checkpoint && (header.extra_data().len() - (SIGNER_VANITY_LENGTH + SIGNER_SIG_LENGTH)) % 20 != 0) {
+			return Err(Box::new(format!("Invalid extra_data length, got {}", header.extra_data().len())).into());
+		}
+
+		// Ensure that the mix digest is zero as we don't have fork protection currently
+		let mixhash = header.decode_seal::<Vec<&[u8]>>()?[0];
+		if mixhash != NULL_MIXHASH {
+			return Err(Box::new("mixhash must be 0x00..0 or 0xff..f.").into());
+		}
+
+		// Ensure that the block doesn't contain any uncles which are meaningless in PoA
+		if *header.uncles_hash() != NULL_UNCLES_HASH {
+			return Err(Box::new(format!(
+				"Invalid uncle hash, got: {}, expected: {}.",
+				header.uncles_hash(),
+				NULL_UNCLES_HASH,
+			)).into());
+		}
+
+		// Ensure that the block's difficulty is meaningful (may not be correct at this point)
+		if *header.difficulty() != DIFF_INTURN && *header.difficulty() != DIFF_NOTURN {
+			return Err(Box::new(format!(
+				"invalid difficulty: expected {} or {}, got: {}.",
+				DIFF_INTURN,
+				DIFF_NOTURN,
+				header.difficulty(),
+			)).into());
+		}
+
+		// All basic checks passed, continue to next phase
+		Ok(())
+	}
+
+	fn verify_block_unordered(&self, _header: &Header) -> Result<(), Error> {
+		// Nothing to check here.
+		Ok(())
+	}
+
+	/// Verify block family by looking up parent state (backfill if needed), then try to apply current header.
+	/// see https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L338
+	fn verify_block_family(&self, header: &Header, parent: &Header) -> Result<(), Error> {
+		// Ignore genesis block.
+		if header.number() == 0 {
+			return Ok(());
+		}
+
+		// parent sanity check
+		if parent.hash() != *header.parent_hash() || header.number() != parent.number() + 1 {
+			return Err(Box::new("invalid parent").into());
+		}
+
+		// Ensure that the block's timestamp isn't too close to it's parent
+		if parent.timestamp() + self.period > header.timestamp() {
+			return Err(Box::new("invalid timestamp").into());
+		}
+
+		// Retrieve the parent state
+		let parent_state = self.state(&parent)?;
+
+		// Try to apply current state, apply() will further check signer and recent signer.
+		let mut new_state = parent_state.clone();
+		new_state.apply(header, header.number() % self.epoch_length == 0)?;
+		new_state.calc_next_timestamp(header, self.period);
+		self.block_state_by_hash.write().insert(header.hash(), new_state);
+
+		Ok(())
+	}
+
+	fn genesis_epoch_data(&self, header: &Header, _call: &Call) -> Result<Vec<u8>, String> {
+		let mut state = self.new_checkpoint_state(header).expect("Unable to parse genesis data.");
+		state.calc_next_timestamp(header, self.period);
+		self.block_state_by_hash.write().insert(header.hash(), state);
+
+		Ok(Vec::new())
```

Comment:
Is it intended to return an empty `Vec` here? If so, please add an inline comment describing why!